Issue #162: Rework configuration format

hsanjuan opened the pull request:
The following commit reimplements ipfs-cluster configuration under
the following premises:

  * Each component is initialized with a configuration object
  defined by its module
  * Each component decides how the JSON representation of its
  configuration looks like
  * Each component parses and validates its own configuration
  * Each component exposes its own defaults
  * Component configurations are make the sections of a
  central JSON configuration file (which replaces the current
  JSON format)
  * Component configurations implement a common interface
  (config.ComponentConfig) with a set of common operations
  * The central configuration file is managed by a
  config.ConfigManager which:
    * Registers ComponentConfigs
    * Assigns the correspondent sections from the JSON file to each
    component and delegates the parsing
    * Delegates the JSON generation for each section
    * Can be notified when the configuration is updated and must be
    saved to disk

The new service.json would then look as follows:

```json
{
  "cluster": {
    "id": "QmTVW8NoRxC5wBhV7WtAYtRn7itipEESfozWN5KmXUQnk2",
    "private_key": "<...>",
    "secret": "00224102ae6aaf94f2606abf69a0e278251ecc1d64815b617ff19d6d2841f786",
    "peers": [],
    "bootstrap": [],
    "leave_on_shutdown": false,
    "listen_multiaddress": "/ip4/0.0.0.0/tcp/9096",
    "state_sync_interval": "1m0s",
    "ipfs_sync_interval": "2m10s",
    "replication_factor": -1,
    "monitor_ping_interval": "15s"
  },
  "consensus": {
    "raft": {
      "heartbeat_timeout": "1s",
      "election_timeout": "1s",
      "commit_timeout": "50ms",
      "max_append_entries": 64,
      "trailing_logs": 10240,
      "snapshot_interval": "2m0s",
      "snapshot_threshold": 8192,
      "leader_lease_timeout": "500ms"
    }
  },
  "api": {
    "restapi": {
      "listen_multiaddress": "/ip4/127.0.0.1/tcp/9094",
      "read_timeout": "30s",
      "read_header_timeout": "5s",
      "write_timeout": "1m0s",
      "idle_timeout": "2m0s"
    }
  },
  "ipfs_connector": {
    "ipfshttp": {
      "proxy_listen_multiaddress": "/ip4/127.0.0.1/tcp/9095",
      "node_multiaddress": "/ip4/127.0.0.1/tcp/5001",
      "connect_swarms_delay": "7s",
      "proxy_read_timeout": "10m0s",
      "proxy_read_header_timeout": "5s",
      "proxy_write_timeout": "10m0s",
      "proxy_idle_timeout": "1m0s"
    }
  },
  "monitor": {
    "monbasic": {
      "check_interval": "15s"
    }
  },
  "informer": {
    "disk": {
      "metric_ttl": "30s",
      "metric_type": "freespace"
    },
    "numpin": {
      "metric_ttl": "10s"
    }
  }
}
```

This new format aims to be easily extensible per component. As such,
it already surfaces quite a few new options which were hardcoded
before.

Additionally, since Go API have changed, some redundant methods have been
removed and small refactoring has happened to take advantage of the new
way.

License: MIT
Signed-off-by: Hector Sanjuan <hector@protocol.ai>

hsanjuan commented:
This was way more difficult than I had anticipated, and in the end took the change to change a lot of small things and do renamings (since API is changing anyway). I still need to do docs and a few tests and fix sharness, but the main code changes should be ready to review.

ZenGround0 commented:
What exactly do you mean that the Go API has changed?  Could you send me a link providing more background on the change?

hsanjuan commented:
@ZenGround0 I mean cluster's Go public methods have changed.

hsanjuan commented:
@ZenGround0 I mean cluster's Go public methods have changed.

hsanjuan commented:
@ZenGround0 I mean cluster's Go public methods have changed.

hsanjuan commented on api/rest/config.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+// Default initializes this Config with working values.
+func (cfg *Config) Default() error {
```
The interface that this Config implements says so. In the case of ipfscluster.Config, Default() means randomly generating keys and using methods which may return errors. I think it doesn't hurt to allow an error to be returned.

hsanjuan commented on api/rest/config.go:
```diff
@@ -0,0 +1,232 @@
+		return err
+	}
+
+	cfg.Default()
```
ack. A further improvement is that only non-zero fields are assigned, in that case, having default for non-assigned fields is necessary. But i'll comment this line and add this as comment

hsanjuan commented on api/rest/config.go:
```diff
@@ -0,0 +1,232 @@
+		// if one is missing, newTLSConfig will
+		// error loudly
+		if !filepath.IsAbs(cert) {
+			cert = filepath.Join(cfg.BaseDir, cert)
```
from config.Saver which is an anonymous field of this Config. config.Saver implements to things which are common to all ComponentConfig stuff and provides BaseDir.

hsanjuan commented on api/rest/config.go:
```diff
@@ -0,0 +1,232 @@
+	jcfg.ReadHeaderTimeout = cfg.ReadHeaderTimeout.String()
+	jcfg.WriteTimeout = cfg.WriteTimeout.String()
+	jcfg.IdleTimeout = cfg.IdleTimeout.String()
+
```
yes!

ZenGround0 commented on consensus/raft/config.go:
```diff
@@ -0,0 +1,212 @@
+	// The unique ID for this server across all time. When running with
+	// ProtocolVersion < 3, you must set this to be the same as the network
+	// address of your transport.
+	// LocalID string `json:local_id`
```
Why is this being commented out?

hsanjuan replied:
It's part of the new Raft version, it won't work right now.

ZenGround0 commented on consensus/raft/logging.go:
```diff
@@ -0,0 +1,35 @@
+package raft
```
A general question: does this refactoring somehow relate to the new config structure, or is it just general cleanup?

hsanjuan replied:
It's general cleanup. These are just moved because they were in the top of the consensus.go file blocking my view of other things.

ZenGround0 commented on consensus/raft/raft.go:
```diff
@@ -302,13 +256,13 @@ func (r *Raft) RemovePeer(peer string) error {
 }
 
 // func (r *Raft) SetPeers(peers []string) error {
-// 	logger.Debugf("SetPeers(): %s", peers)
```
Should SetPeers be removed?

hsanjuan replied:
I'm not sure, probably when I deal with the new Raft version.

ZenGround0 commented on docs/ipfs-cluster-guide.md:
```diff
@@ -56,23 +56,91 @@ You can deploy cluster in the way which fits you best, as both ipfs and ipfs-clu
-The configuration stores a `cluster_secret` which is a 32 byte (hex-encoded) key which **must be shared by all cluster peers**. Using an empty key has security implications (see the Security section). Using different keys will prevent different peers from talking to each other.
-
-The configuration options are documented at: https://godoc.org/github.com/ipfs/ipfs-cluster#JSONConfig . It is recommended that you read them carefully before running your cluster.
+The ipfs-cluster configuration file is usually found at `~/.ipfs-cluster/service.json`. It holfd all the configurable options for cluster and its different components.
```
Typo: "It holfd all the... "

ZenGround0 commented on docs/ipfs-cluster-guide.md:
```diff
@@ -56,23 +56,91 @@ You can deploy cluster in the way which fits you best, as both ipfs and ipfs-clu
 ## Starting your cluster
 
-`ipfs-cluster-service` will launch your cluster peer. If you have not configured any `cluster_peers` in the configuration, nor any `bootstrap` addresses, a single-peer cluster will be launched.
+`ipfs-cluster-service` will launch your cluster peer. If you have not configured any `cluster/peers` in the configuration, nor any `cluster/bootstrap` addresses, a single-peer cluster will be launched.
```
It might be more json-idiomatic to instead write `cluster.peers` and `cluster.bootstrap`

hsanjuan replied:
changed all occurrences.

ZenGround0 commented on docs/ipfs-cluster-guide.md:
```diff
@@ -91,9 +159,9 @@ For example, a commit operation to the log is triggered with  `ipfs-cluster-ctl
 The "peer add" and "peer remove" operations also trigger log entries and fully depend on a healthy consensus status. Modifying the cluster peers is a tricky operation because it requires informing every peer of the new peer set. If a peer is down during this operation, it is likely that it will not learn about it when it comes up again. Thus, it is recommended to bootstrap the peer again when the ...
 
-The consensus log data is backed in the `consensus_data_folder`. This folder stores two types of information: the [boltDB] database storing the Raft log, and the state snapshots. Snapshots from the log are performed regularly when the log grows too big (or on shutdown). When a peer is far behind in catching up with the log, Raft may opt to send a snapshot directly, rather than to send every log e ...
+The consensus log data is backed in the `ipfs-cluster-data` subfolder, next to the main configuration file. This folder stores two types of information: the [boltDB] database storing the Raft log, and the state snapshots. Snapshots from the log are performed regularly when the log grows too big (or on shutdown). When a peer is far behind in catching up with the log, Raft may opt to send a snapsho ...
```
As this is configurable, perhaps we could say "By default, the consensus log data is ..."

hsanjuan replied:
fixed

ZenGround0 commented on docs/ipfs-cluster-guide.md:
```diff
@@ -118,18 +186,19 @@ In normal operation, all three states are in sync, as updates to the *shared sta
 ## Static cluster membership considerations
 
-We call a static cluster, that in which the set of `cluster_peers` is fixed, where "peer add"/"peer rm"/bootstrapping operations don't happen (at least normally) and where every cluster member is expected to be running all the time.
+We call a static cluster, that in which the set of `cluster/peers` is fixed, where "peer add"/"peer rm"/bootstrapping operations don't happen (at least normally) and where every cluster member is expected to be running all the time.
```
same as 137

ZenGround0 commented on docs/ipfs-cluster-guide.md:
```diff
@@ -118,18 +186,19 @@ In normal operation, all three states are in sync, as updates to the *shared sta
 ## Dynamic cluster membership considerations
 
-We call a dynamic cluster, that in which the set of `cluster_peers` changes. Nodes are bootstrapped to existing cluster peers, the "peer add" and "peer rm" operations are used and/or the `leave_on_shutdown` configuration option is enabled. This option allows a node to abandon the consensus membership when shutting down. Thus reducing the cluster size by one.
+We call a dynamic cluster, that in which the set of `cluster/peers` changes. Nodes are bootstrapped to existing cluster peers, the "peer add" and "peer rm" operations are used and/or the `cluster/leave_on_shutdown` configuration option is enabled. This option allows a node to abandon the consensus membership when shutting down. Thus reducing the cluster size by one.
```
same as 137

ZenGround0 commented on docs/ipfs-cluster-guide.md:
```diff
@@ -118,18 +186,19 @@ In normal operation, all three states are in sync, as updates to the *shared sta
 
 Dynamic clusters allow greater flexibility at the cost of stablity. Join and leave operations are tricky as they change the consensus membership and they are likely to create bad situations in unhealthy clusters. Also, bear in mind than removing a peer from the cluster will trigger a re-allocation of the pins that were associated to it. If the replication factor was 1, it is recommended to keep t ...
 
+Peers joining an existing cluster should have a non-divergent state. That means: their consensus `ipfs-cluster-data` folder should be empty or, if not, it should contain data belonging to the existing cluster. A joining peer should have not been running on it's own separately from the running cluster. When a peer leaves or is removed, any existing peers will be saved as `bootstrap` peers, so that ...
```
Typo:  should be "on its own" (no apostrophe).  Thanks for making our bootstrap save on removal explicit

hsanjuan replied:
Fixed. Thanks for finding all these typos!

ZenGround0 commented on informer/disk/disk.go:
```diff
@@ -14,20 +11,18 @@ import (
 	"github.com/ipfs/ipfs-cluster/api"
 )
 
+// MetricType identifies the type of metric to fetch from the IPFS daemon.
```
#Documentation Friday

ZenGround0 commented on informer/numpin/config.go:
```diff
@@ -0,0 +1,72 @@
+	return configKey
+}
+
+// Default initializes this Config with sensisble values.
```
Typo: "Sensisble"

hsanjuan replied:
fixed

hsanjuan commented on cluster_config.go:
```diff
@@ -0,0 +1,382 @@
+		return errors.New("cluster.monitoring_interval is invalid")
+	}
+	return nil
+}
```
fixed

hsanjuan commented on cluster_config.go:
```diff
@@ -0,0 +1,382 @@
+	err = json.Unmarshal(raw, jcfg)
+	if err != nil {
+		logger.Error("Error unmarshaling cluster config")
+		return err
```
true. actually I'm not a fan of implicit return variables. I have updated it.

hsanjuan commented on cluster_config.go:
```diff
@@ -0,0 +1,382 @@
+	}
+
+	// Bootstrap peers
+	bootstrap := make([]string, len(cfg.Bootstrap), len(cfg.Bootstrap))
```
My feeling is that it is not a bad practice to use capacity in things which will not/should not exceed that capacity. Agreed it's not super important here.

hsanjuan commented on cluster_config.go:
```diff
@@ -0,0 +1,382 @@
+}
+
+// EncodeClusterSecret converts a byte slice to its hex string representation.
+func EncodeClusterSecret(secretBytes []byte) string {
```
They are convenience functions for the user. I think ipfs-cluster-service uses DecodeClusterSecret when it reads it interactively for example.

hsanjuan commented on config/config.go:
```diff
@@ -0,0 +1,343 @@
+
+// Section is a section of which stores
+// component-specific configurations.
+type Section map[string]ComponentConfig
```
To question 1:
* Might think of a future when you can choose which consensus component to use at launch, and still keep configurations for both modules in the configuration. Also, closer to reality is that we think of having several Informer components and each has a configuration.

To question 2: Fair enough, I have changed it.

hsanjuan commented on config/config.go:
```diff
@@ -0,0 +1,343 @@
+}
+
+// RegisterComponent lets the Manager load and save component configurations
+func (cfg *Manager) RegisterComponent(t ComponentType, ccfg ComponentConfig) error {
```
yeah you're right. Fixed

hsanjuan replied:
I mean, could always decide to return errors upon re-registration of a component or something, but we can change it then.

hsanjuan commented on config/config.go:
```diff
@@ -0,0 +1,343 @@
+
+	cfg.sections[t][ccfg.ConfigKey()] = ccfg
+	if t == Cluster { // this is a special section
+		cfg.clusterKey = key
```
That's correct. Unlike the others, there can be only one cluster config in the JSON so cfg.clusterKey keeps track of which one is the "one".

It seemed like it made sense to internally consider the cluster section just as the others for looping them easier, but maybe it's clearer to have a clusterSection directly in the Manager object. Any preference?

ZenGround0 replied:
My preference is to have a clusterSection explicitly in the Manager.  That said it's not a crazy strong preference; if you prefer this way I am good with keeping it as is.

hsanjuan commented on config/config.go:
```diff
@@ -0,0 +1,343 @@
+	return err
+}
+
+// LoadJSON reads parses configurations for all registered components,
```
fixed

hsanjuan commented on config/config.go:
```diff
@@ -0,0 +1,343 @@
+
+	// Load Cluster section. Needs to have been registered
+	clustercfgs := cfg.sections[Cluster]
+	if clustercfgs != nil && len(clustercfgs) > 0 {
```
yeah. would you prevent it?

ZenGround0 replied:
If you decide not to go with an explicit clusterSection in the Manager I think we should do something if `len(clustercfgs) > 1`, maybe an error, maybe a warning, because that would indicate that the user did not understand the invariant that there can only be one cluster section.

hsanjuan commented on config/config.go:
```diff
@@ -0,0 +1,343 @@
+	}
+
+	// Helper function to load json from each section in the json config
+	loadCompJSON := func(section Section, jsonSection jsonSection) error {
```
done

hsanjuan commented on config/config.go:
```diff
@@ -0,0 +1,343 @@
+
+	// Given a Section and a jsonSection, it updates the
+	// components configurations in the latter if the components
+	// configurations in the former have changed.
```
Sorry, that comment was outdated. It updates component configurations no matter what.

hsanjuan commented:
I have rebased on master and addressed another bunch of comments. Keeping addressing of comments on separate commits so it can be checked easier. Will squash at the end.

ZenGround0 commented on sharness/lib/test-lib.sh:
```diff
@@ -27,7 +27,7 @@ test_ipfs_init() {
         echo "Docker not found"
         exit 1
     fi
-    if docker ps -a --format '{{.Names}}' | egrep -q '^ipfs$'; then
```
Why did you drop the "-a"?  I didn't research the implications of adding it fully, but I added this in the latest fix-sharness PR because docker ps without "-a" couldn't find my running ipfs container and kept causing a tricky crash when running sharness several times in a row.

hsanjuan replied:
Yeah I thought it was no big deal then. The -a means  "Show all containers (default shows just running)". But debugging myself such error I was getting the "ipfs container already running" message when the ipfs container was not in running state (there was a stuck one in Created state I think, probably same issue you had). which was not accurate.

I'm going to open an issue about hardening this shell script a bit. With our without -a this should attempt to kill/remove existing containers with that name or log more info.

hsanjuan commented on ipfs-cluster-service/dist/README.md:
````diff
@@ -31,42 +31,19 @@ All peers in a cluster **must share the same cluster secret**. Using an empty se
-    "allocation_strategy": "reposize
-}
-```
+You can add the multiaddresses for the other cluster peers to the `cluster/peers` or `cluster/bootstrap` variables (see below). A configuration example with explanations is provided in [A guide to running IPFS Cluster](https://github.com/ipfs/ipfs-cluster/blob/master/docs/ipfs-cluster-guide.md).
````
fixed

hsanjuan commented on ipfs-cluster-service/dist/README.md:
```diff
@@ -31,42 +31,19 @@ All peers in a cluster **must share the same cluster secret**. Using an empty se
+The configuration file should probably be identical among all cluster peers, except for the `id` and `private_key` fields. Once every cluster peer has the configuration in place, you can run `ipfs-cluster-service` to start the cluster.
 
-#### Clusters using `cluster_peers`
+#### Clusters using `cluster/peers`
```
fixed

hsanjuan commented on ipfs-cluster-service/dist/README.md:
```diff
@@ -31,42 +31,19 @@ All peers in a cluster **must share the same cluster secret**. Using an empty se
+The `peers` configuration variable holds a list of current cluster members. If you know the members of the cluster in advance, or you want to start a cluster fully in parallel, set `peers` in all configurations so that every peer knows the rest upon boot. Leave `bootstrap` empty. A cluster peer address looks like: `/ip4/1.2.3.4/tcp/9096/<id>`.
 
-#### Clusters using `bootstrap`
+#### Clusters using `cluster/bootstrap`
```
fixed

hsanjuan commented:
More comments addressed.

I have added a `clusterConfig` key to the configuration object.

I am missing a bunch of tests at least to bring up the code coverage back up to the previous levels. Will work on that tomorrow.

ZenGround0 commented:
@hsanjuan Other than the lack of tests you mentioned above this LGTM.

ZenGround0 commented:
@hsanjuan I am submitting approval under the assumption the last commit just squashes the previous commits and doesn't include changes.  Let me know if this is incorrect.  Thanks

hsanjuan commented:
That's correct (it did fix a small thing too in config.Default())

hsanjuan commented on api/rest/config_test.go:
```diff
@@ -0,0 +1,92 @@
+		t.Fatal("error validating")
+	}
+
+	cfg.ListenAddr = nil
```
I'd say I'm rather testing Default, and the way to do it is validating default configurations. Gonna leave it like this, since Validate is also tested separately.

hsanjuan commented on monitor/basic/config_test.go:
```diff
@@ -0,0 +1,52 @@
+func TestToJSON(t *testing.T) {
+	cfg := &Config{}
+	cfg.LoadJSON(cfgJSON)
+	_, err := cfg.ToJSON()
```
Cannot match directly against cfgJSON because indentation or key order might be changed, but it's probably a good thing to re-parse the generated json and check it validates.
